pkg/providers/cloudstack/validator.go

```diff
@@ -244,6 +245,23 @@ func (v *Validator) validateMachineConfig(ctx context.Context, datacenterConfig
 		if err = v.cmk.ValidateServiceOfferingPresent(ctx, zone.Id, machineConfig.Spec.ComputeOffering); err != nil {
 			return fmt.Errorf("validating service offering: %v", err)
 		}
+		if len(machineConfig.Spec.DiskOffering.Id) > 0 || len(machineConfig.Spec.DiskOffering.Name) > 0 {
+			if !machineConfig.Spec.DiskOffering.ValidatePath() {
+				return fmt.Errorf("mountPath: %s invalid, must be non-empty and starts with /", machineConfig.Spec.DiskOffering.MountPath)
+			}
+			if len(machineConfig.Spec.DiskOffering.Device) == 0 {
+				return fmt.Errorf("device: %s invalid, must be non-empty", machineConfig.Spec.DiskOffering.Device)
+			}
+			if len(machineConfig.Spec.DiskOffering.Filesystem) == 0 {
+				return fmt.Errorf("filesystem: %s invalid, must be non-empty", machineConfig.Spec.DiskOffering.Filesystem)
+			}
+			if len(machineConfig.Spec.DiskOffering.Label) == 0 {
```

Comment:
aren't these required fields already validated the type definition?

Reply:
If you provided the required fields with an empty string in the yaml file, it can pass the type definition and I think this validator is still needed.